Add new types of document transformers

jasonwcfan opened the pull request:
- Description: Add two new document transformers that translates documents into different languages and converts documents into q&a format to improve vector search results. Uses OpenAI function calling via the [doctran](https://github.com/psychic-api/doctran/tree/main) library.
  - Issue: N/A
  - Dependencies: `doctran = "^0.0.5"`
  - Tag maintainer: @rlancemartin @eyurtsev @hwchase17 
  - Twitter handle: @psychicapi or @jfan001

Notes
- Adheres to the `DocumentTransformer` abstraction set by @dev2049 in #3182 
- refactored `EmbeddingsRedundantFilter` to put it in a file under a new `document_transformers` module
- Added basic docs for `DocumentInterrogator`, `DocumentTransformer` as well as the existing `EmbeddingsRedundantFilter`

jasonwcfan committed:
add interrogator and translator document transformers

jasonwcfan committed:
add as python dependency

jasonwcfan committed:
fix doc formatting

rlancemartin reviewed (commented):
Great! We should add a `notebook` e.g., `doctran.ipynb` that shows usage similar to the one below:

https://github.com/hwchase17/langchain/blob/master/docs/extras/modules/data_connection/document_loaders/integrations/source_code.ipynb

Cristobal got very good engagement on his related thread, and there's a lot of interest in this theme:

https://twitter.com/cristobal_dev/status/1675745314592915456?s=20

rlancemartin commented on langchain/document_transformers/interrogate.py:
```diff
@@ -0,0 +1,32 @@
+from doctran import Doctran
+
+
+class DocumentInterrogator(BaseDocumentTransformer, BaseModel):
```
We might consider making this a `BaseBlobParser` similar to what we do [here](https://github.com/hwchase17/langchain/blob/705d2f5b9225e63ad4603776d5047d18a6c3fdfa/langchain/document_loaders/parsers/grobid.py#L14C20-L14C34).

And then it is called easily w/o any boilerplate loader code:

```
loader = GenericLoader.from_filesystem(
    "./example_data/pds",
    glob="*",
    suffixes=[".pdf"],
    parser=DocumentTranslator()
)
docs = loader.load()
```

jasonwcfan replied:
Sounds good I'll take a look at adding that. These transformers depend on an OpenAI call so need to be run async - would that cause any issues when used as a parser?

rlancemartin replied:
No, we have a similar parser that uses OpenAI Whisper.

jasonwcfan committed:
add notebook examples

jasonwcfan commented:
Added the notebook examples - let me know if it's good to merge @rlancemartin and I can add the blob stuff in a followup PR, or if you want me to include it into this one!

rlancemartin commented on langchain/document_transformers/interrogate.py:
```diff
@@ -0,0 +1,32 @@
+from doctran import Doctran
+
+
+class DocumentInterrogator(BaseDocumentTransformer, BaseModel):
```
Maybe we call this `QA_generator` as a more descriptive name?

rlancemartin replied:
It should be minimal changes to make this a `parser`. 

1/ Inherit from `BaseBlobParser`
2/ Just re-name `atransform_documents` to `lazy_parse` (as shown [here](https://github.com/hwchase17/langchain/blob/2eb4a2ceeab8fb2e7b64461c71e10c468ff0ca01/langchain/document_loaders/parsers/audio.py#L15C5-L15C8)) to accept blobs
3/ should then work out-of-the-box w/ the blob loaders

The UX in the notebook is then simple:

```
loader = GenericLoader.from_filesystem(
    "./example_data/pdf",
    glob="*",
    suffixes=[".pdf"],
    parser= QA_generator()
)
docs = loader.load()
```

rlancemartin commented on langchain/document_transformers/interrogate.py:
```diff
@@ -0,0 +1,32 @@
+        self, documents: Sequence[Document], **kwargs: Any
+    ) -> Sequence[Document]:
+        """Asynchronously convert documents into Q&A format with doctran and OpenAI functions."""
+        openai_api_key = get_from_dict_or_env(kwargs, "openai_api_key", "OPENAI_API_KEY")
```
Can make api_key optional, as done [here](https://github.com/hwchase17/langchain/blob/2eb4a2ceeab8fb2e7b64461c71e10c468ff0ca01/langchain/document_loaders/parsers/audio.py#L13).

rlancemartin commented:
> Added the notebook examples - let me know if it's good to merge @rlancemartin and I can add the blob stuff in a followup PR, or if you want me to include it into this one!

Thanks! 

> Added the notebook examples - let me know if it's good to merge @rlancemartin and I can add the blob stuff in a followup PR, or if you want me to include it into this one!

Great! Just added comments. 

IMO, easier to make it a parser. Minimal code changes. I can help if needed.

Will simplify UX and be consistent w/ other recent transformers.

rlancemartin commented:
Also, functionality here is:
1/ QA
2/ Translate

Looks like Doctran can [do more](https://github.com/psychic-api/doctran/tree/main) in terms of metadata extraction.

Plans to add that in a follow-up? We can discuss, too.

QA generation is cool, but impact is somewhat mild b/c we already have QAGeneration chain. 

Translation is neat. 

Adding context to chunks (in metadata) is the high impact addition that would be great to get in.

jasonwcfan commented:
@rlancemartin ack, will add this in as a parser tomorrow.

Re: naming I avoided including `QA` because there's already a lot of modules in LangChain that have some variation of "QA" in its name so I thought it might just increase confusion - but happy to call it whatever you think makes sense!

Re: additional functionality yeah I can add those too in a follow up, we can discuss on monday to see which ones are most useful and the best way to include them. Would also be curious to understand the vision is for the document transformer module, I assumed it would be for these kinds of transformations

jasonwcfan committed:
add extract, qa, translate parsers

jasonwcfan commented:
Added translate and QA as parsers, as well as metadata extraction since you mentioned that was valuable. 

I ran into a couple of issues though:

**Passing arguments to parsers** 
The `BaseBlobParser` class doesn't support additional arguments when passed into loaders. It seems that other parsers don't need to accept additional arguments, but it's necessary for translation (need to specify the language) and metadata extraction (need to pass in a JSON schema to describe the metadata to extract). 

I can extend `BaseBlobParser` and `BaseLoader` to accept **kwargs if we want to keep these modules as parsers, but that might make the scope of these changes further reaching than anticipated.

```python
class DoctranTranslateParser(BaseBlobParser):
    """Translates text documents into other languages using doctran."""

    def lazy_parse(self, blob: Blob, language: str, **kwargs) -> Iterator[Document]:
        """Lazily parse the blob."""
        openai_api_key = get_from_dict_or_env(kwargs, "openai_api_key", "OPENAI_API_KEY")
        doctran = Doctran(openai_api_key=openai_api_key)
        doctran_doc = doctran.parse(content=blob.as_string()).translate(language=language).execute()
        yield [Document(page_content=doctran_doc.properties_as_yaml() if doctran_doc.extracted_properties else doctran_doc.transformed_content)]
```

**Splitting**
When splitting documents, the metadata for the parent document is copied to all child documents. This means any extracted properties (including QA pairs) will be copied over to all chunks, rather than mapped to only the relevant chunks that contain those properties. Therefore vector retrieval using these metadata properties won't be able to retrieve the correct documents if the document gets split.

I can:
1. Keep the functionality as is, but that would make metadata extraction useless if the user intends to split the document after parsing.
2. Keep the metadata extraction/QA generation step as a document transformation so it can happen after splitting
3. Add functionality to handle document splitting while mapping properties to the correct chunks. Not sure if there's a great way to do this since I'm not super familiar with the LangChain codebase yet.

Let me know what you think! Can also discuss more on our call but it's a lot to cover in 30 min haha.

jasonwcfan committed:
make openai_api_key optional parameter in the parser constructor

rlancemartin commented:
> Added translate and QA as parsers, as well as metadata extraction since you mentioned that was valuable.  Will clean up the old files and update the notebooks early this week.
> 
> 
> 
> I ran into a couple of issues though:
> 
> 
> 
> **Passing arguments to parsers** 
> 
> The `BaseBlobParser` class doesn't support additional arguments when passed into loaders. It seems that other parsers don't need to accept additional arguments, but it's necessary for translation (need to specify the language) and metadata extraction (need to pass in a JSON schema to describe the metadata to extract). 
> 
> 
> 
> I can extend `BaseBlobParser` and `BaseLoader` to accept **kwargs if we want to keep these modules as parsers, but that might make the scope of these changes further reaching than anticipated.
> 
> 
> 
> ```python
> 
> class DoctranTranslateParser(BaseBlobParser):
> 
>     """Translates text documents into other languages using doctran."""
> 
> 
> 
>     def lazy_parse(self, blob: Blob, language: str, **kwargs) -> Iterator[Document]:
> 
>         """Lazily parse the blob."""
> 
>         openai_api_key = get_from_dict_or_env(kwargs, "openai_api_key", "OPENAI_API_KEY")
> 
>         doctran = Doctran(openai_api_key=openai_api_key)
> 
>         doctran_doc = doctran.parse(content=blob.as_string()).translate(language=language).execute()
> 
>         yield [Document(page_content=doctran_doc.properties_as_yaml() if doctran_doc.extracted_properties else doctran_doc.transformed_content)]
> 
> ```
> 
> 
> 
> **Splitting**
> 
> When splitting documents, the metadata for the parent document is copied to all child documents. This means any extracted properties (including QA pairs) will be copied over to all chunks, rather than mapped to only the relevant chunks that contain those properties. Therefore vector retrieval using these metadata properties won't be able to retrieve the correct documents if the document gets split.
> 
> 
> 
> I can:
> 
> 1. Keep the functionality as is, but that would make metadata extraction useless if the user intends to split the document after parsing.
> 
> 2. Keep the metadata extraction/QA generation step as a document transformation so it can happen after splitting
> 
> 3. Add functionality to handle document splitting while mapping properties to the correct chunks. Not sure if there's a great way to do this since I'm not super familiar with the LangChain codebase yet.
> 
> 
> 
> Let me know what you think! Can also discuss more on our call but it's a lot to cover in 30 min haha.

Thanks! Hmm, you should be able to pass args to the parser. I'll have a close look later today!

'''

loader = GenericLoader.from_filesystem(
    "../Papers/",
    glob="*",
    suffixes=[".pdf"],
    parser= GrobidParser(segment_sentences=False)
)
docs = loader.load()
'''

jasonwcfan commented:
re: arguments, ah yeah nevermind - didn't notice that the loader accepts an instance of a parser rather than the class, so I can just include set them in the constructor 👍

jasonwcfan committed:
move params from lazy_parse to constructor

rlancemartin commented:
> re: arguments, ah yeah nevermind - didn't notice that the loader accepts an instance of a parser rather than the class, so I can just include set them in the constructor 👍 

Nice! Ya. And I think chunk-wise metadata should work (let's have a quick skim at Grobid as an example). I'll look when back on my computer later.

rlancemartin commented on pyproject.toml:
```diff
@@ -116,6 +116,7 @@ streamlit = {version = "^1.18.0", optional = true, python = ">=3.8.1,<3.9.7 || >
 psychicapi = {version = "^0.8.0", optional = true}
 cassio = {version = "^0.0.7", optional = true}
 rdflib = {version = "^6.3.2", optional = true}
+doctran = {version = "^0.0.5", optional = true}
```
Just include this in the notebook rather than adding to `pyproject.toml`

```
! pip install doctran
```

jasonwcfan replied:
How does it get loaded in the parser modules if it's not included as a dependency?

rlancemartin commented on docs/extras/modules/data_connection/document_transformers/interrogate_document.ipynb:
```diff
@@ -0,0 +1,223 @@
+   "outputs": [],
+   "source": [
+    "from langchain.schema import Document\n",
+    "from langchain.document_transformers import DocumentInterrogator"
```
Running notebook, I see import error:

```
│    6 from langchain.utils import get_from_dict_or_env                                            │
│    7 from langchain.schema import BaseDocumentTransformer, Document                              │
│ ❱  8 from doctran import Doctran                                                                 │
│    9                                                                                             │
│   10                                                                                             │
│   11 class DocumentInterrogator(BaseDocumentTransformer, BaseModel):                             │
│                                                                                                  │
│ /Users/rlm/anaconda3/envs/lcn2/lib/python3.9/site-packages/doctran/__init__.py:1 in <module>     │
│                                                                                                  │
│ ❱ 1 from .doctran import Doctran, Document, DoctranConfig, ContentType, ExtractProperty, Rec     │
│   2                                                                                              │
│                                                                                                  │
│ /Users/rlm/anaconda3/envs/lcn2/lib/python3.9/site-packages/doctran/doctran.py:14 in <module>     │
│                                                                                                  │
│    11 from typing import List, Optional, Dict, Any, Literal                                      │
│    12 from pydantic import BaseModel                                                             │
│    13                                                                                            │
│ ❱  14 class ExtractProperty(BaseModel):                                                          │
│    15 │   name: str                                                                              │
│    16 │   description: str                                                                       │
│    17 │   type: Literal["string", "number", "boolean", "array", "object"]                        │
│                                                                                                  │
│ /Users/rlm/anaconda3/envs/lcn2/lib/python3.9/site-packages/doctran/doctran.py:18 in              │
│ ExtractProperty                                                                                  │
│                                                                                                  │
│    15 │   name: str                                                                              │
│    16 │   description: str                                                                       │
│    17 │   type: Literal["string", "number", "boolean", "array", "object"]                        │
│ ❱  18 │   items: Optional[List | Dict[str, Any]]                                                 │
│    19 │   enum: Optional[List[str]]                                                              │
│    20 │   required: bool = True                                                                  │
│    21                            
```

jasonwcfan commented:
@rlancemartin I might be missing something here, but is there an easy way to load strings into Blobs? Right now I have to define a new BlobLoader class in order to load strings:

```python
class StringBlobLoader(BlobLoader):
    def yield_blobs(self) -> Iterable[Blob]:
        yield Blob(data=sample_text)

loader = GenericLoader(
    blob_loader=StringBlobLoader(),
    blob_parser= DoctranTranslateParser(language="spanish")
)
docs = loader.load()
```

rlancemartin commented:
> @rlancemartin I might be missing something here, but is there an easy way to load strings into Blobs? Right now I have to define a new BlobLoader class in order to load strings:
> 
> ```python
> class StringBlobLoader(BlobLoader):
>     def yield_blobs(self) -> Iterable[Blob]:
>         yield Blob(data=sample_text)
> 
> loader = GenericLoader(
>     blob_loader=StringBlobLoader(),
>     blob_parser= DoctranTranslateParser(language="spanish")
> )
> docs = loader.load()
> ```

Blobs are files, so you can just put an example txt file w/ your string in [`../example data`](https://github.com/hwchase17/langchain/tree/master/docs/extras/modules/data_connection/document_loaders/integrations/example_data).

jasonwcfan committed:
add notebook examples and refactor to be parsers

jasonwcfan commented:
@rlancemartin Good to know! It wasn't a large amount of data so I just kept it in the notebook :)

All the changes are in!
* fixed the bug that was causing the issue in python <3.10, turns out it was due to the new Union notation using `|`.
* removed doctran as a dependency but will raise an `ImportError` if these new parsers are used without the package being installed

I think it should be good to merge in now, but let me know if I missed anything!

jasonwcfan committed:
Merge branch 'master' into master

rlancemartin commented:
> @rlancemartin Good to know! It wasn't a large amount of data so I just kept it in the notebook :)
> 
> 
> 
> All the changes are in!
> 
> * fixed the bug that was causing the issue in python <3.10, turns out it was due to the new Union notation using `|`.
> 
> * removed doctran as a dependency but will raise an `ImportError` if these new parsers are used without the package being installed
> 
> 
> 
> I think it should be good to merge in now, but let me know if I missed anything!

Nice! I'll have a look tmrw and plan to get it in!

rlancemartin commented on docs/extras/modules/data_connection/document_transformers/interrogate_document.ipynb:
```diff
@@ -0,0 +1,246 @@
+    "        \n",
+    "loader = GenericLoader(\n",
+    "    blob_loader=StringBlobLoader(),\n",
+    "    blob_parser= DoctranQAParser(openai_api_key=\"YOUR_OPENAI_API_KEY\") # openai_api_key can be excluded if the OPENAI_API_KEY environment variable is set.\n",
```
See an odd error running; OpenAI API key is an env var?
```
Exception: Error executing transformation (<Transformation.interrogate: 'DocumentInterrogator'>, {}): OpenAI 
function call failed: Incorrect API key provided: YOUR_OPE*******_KEY. You can find your API key at 
https://platform.openai.com/account/api-keys.
```

jasonwcfan replied:
Fixed - removed that and updated the parsers to set the openai_api_key in the constructor, so now if the env variable is not set it gives the standard message to set the env variable or pass it as a parameter.

rlancemartin commented on langchain/document_loaders/parsers/text_translate.py:
```diff
@@ -0,0 +1,29 @@
+            openai_api_key = self.openai_api_key
+        else:
+            openai_api_key = get_from_env("openai_api_key", "OPENAI_API_KEY")
+        doctran_doc = self.doctran.parse(content=blob.as_string()).translate(language=self.language).execute()
```
I think you need to pass `openai_api_key` here.

jasonwcfan replied:
Same fix as above, openai_api_key is passed into doctran constructor so I handled this all in the constructor for the parser

rlancemartin commented:
Also, see tests:
```
poetry run mypy .
tests/integration_tests/test_document_transformers.py:2: error: Module "langchain.document_transformers" has no attribute "EmbeddingsClusteringFilter"  [attr-defined]
```

jasonwcfan committed:
remove api key

jasonwcfan committed:
updated openai api key handling

jasonwcfan committed:
add import for EmbeddingsClusteringFilter

jasonwcfan committed:
add additional import

jasonwcfan commented:
Fixed the test failure as well! It was a result of moving `document_transformers.py` into a directory with an `__init__.py` Thanks for catching these @rlancemartin

rlancemartin commented:
> Fixed the test failure as well! It was a result of moving `document_transformers.py` into a directory with an `__init__.py` Thanks for catching these @rlancemartin

hmm, pulled your latest but still see issues:
```
File ~/anaconda3/envs/lcn2/lib/python3.9/site-packages/doctran/doctran.py:165, in Doctran.__init__(self, openai_api_key, openai_model, openai_token_limit)
    164 def __init__(self, openai_api_key: str = None, openai_model: str = "gpt-4", openai_token_limit: int = 8000):
--> 165     self.config = DoctranConfig(
    166         openai_api_key=openai_api_key,
    167         openai_model=openai_model,
    168         openai_completions_url="https://api.openai.com/v1/completions",
    169         openai=openai,
    170         openai_token_limit=openai_token_limit
    171     )
    172     if openai_api_key:
    173         self.config.openai.api_key = openai_api_key

File ~/anaconda3/envs/lcn2/lib/python3.9/site-packages/pydantic/main.py:341, in pydantic.main.BaseModel.__init__()

ValidationError: 1 validation error for DoctranConfig
openai_api_key
  none is not an allowed value (type=type_error.none.not_allowed)
```

jasonwcfan committed:
fix bug where self.openai_api_key was missing

jasonwcfan commented:
@rlancemartin sorry about that I was running the notebooks with the API key passed in as a param so wasn't running into that bug, but tested it with the key set as an environment variable and fixed the issue.

rlancemartin commented:
> @rlancemartin sorry about that I was running the notebooks with the API key passed in as a param so wasn't running into that bug, but tested it with the key set as an environment variable and fixed the issue.

Cool, Ill give it one more check later.

In the meantime, see failing test:
```
langchain/document_transformers/__init__.py:1: error: Module "langchain.document_transformers.embeddings_redundant_filter" has no attribute "_filter_cluster_embeddingsgit"; maybe "_filter_cluster_embeddings" or "_filter_similar_embeddings"?  [attr-defined]
Found 1 error in 1 file (checked 1332 source files)
make: *** [Makefile:34: lint] Error 1
```

jasonwcfan committed:
remove typo

jasonwcfan committed:
Merge branch 'master' into master

rlancemartin committed:
Move to document transformer per discussion w/ @baskaryan

rlancemartin commented:
@jasonwcfan i chatted w/ @baskaryan. this is better as a `document transformer` since it clearly ingests a document (rather than loading from a novel source) and transforms it. i moved it back to being a `transformer`. please have a quick look, and then we can merge tomorrow. please add a bit more detail to `Extract Properties` notebook header and also make sure that part is working as expected. it looks a bit off to me.

rlancemartin commented:
Note: I'm using `poetry lock --no-update` to re-generate `poetry.lock` after checking out `master` `pyproject.toml`. Good for future reference when you see conflicts w/ poetry.lock.

rlancemartin committed:
Fix poetry.lock

rlancemartin committed:
Fix poetry.lock

rlancemartin commented:
Also we should have a careful look at what's going on w/ API keys. My API key got disable from using Doctran, I believe! I wasn't doing anything else this evening. ![image](https://github.com/hwchase17/langchain/assets/122662504/4f462bd2-7f35-440a-a331-7599a691b8c3)

jasonwcfan committed:
update notebooks, __init__.py, and transformers

jasonwcfan commented:
Oof that's odd, I can't think of any way it would have leaked through Doctran. It's not in the repo and we don't send openai keys to any API. I'll comb through doctran tomorrow and double check.

In the meantime I've pushed some updates to the notebooks + transformers, I'll check if they pass the tests once it's done deploying. Thanks for bearing with me @rlancemartin!

rlancemartin commented:
> Oof that's odd, I can't think of any way it would have leaked through Doctran. It's not in the repo and we don't send openai keys to any API. I'll comb through doctran tomorrow and double check.
> 
> 
> 
> In the meantime I've pushed some updates to the notebooks + transformers, I'll check if they pass the tests once it's done deploying. Thanks for bearing with me @rlancemartin!

Sounds good! Ya it's really weird.

rlancemartin commented:
Also, error related to use of `await` testing `extract_properties` notebook -

```
 in <module>:28                                                                                   │
│                                                                                                  │
│   25 │   }                                                                                       │
│   26 ]                                                                                           │
│   27                                                                                             │
│ ❱ 28 extracted_document = await property_extractor.atransform_documents(documents,properties=    │
│   29                                                                                             │
│                                                                                                  │
│ /Users/rlm/Desktop/Code/langchain/langchain/document_transformers/text_extract.py:38 in          │
│ atransform_documents                                                                             │
│                                                                                                  │
│   35 │   │   │   )                                                                               │
│   36 │   │   properties = [ExtractProperty(**property) for property in properties]               │
│   37 │   │   for d in documents:                                                                 │
│ ❱ 38 │   │   │   doctran_doc = await (                                                           │
│   39 │   │   │   │   doctran.parse(content=d.page_content)                                       │
│   40 │   │   │   │   .extract(properties=properties)                                             │
│   41 │   │   │   │   .execute()                                                                  │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
TypeError: object Document can't be used in 'await' expression
```

rlancemartin committed:
fmt and test

rlancemartin committed:
Update poetry.lock

rlancemartin committed:
Merge master

rlancemartin committed:
Update poetry.lock after merge

rlancemartin committed:
Fmt

rlancemartin committed:
Update notebooks w/ pip cmd and fmt

rlancemartin commented:
Adding @baskaryan re: creation of `document_transformers` subdir and moving [document_transformers.py ](https://github.com/hwchase17/langchain/blob/master/langchain/document_transformers.py) to `langchain/document_transformers/embeddings_redundant_filter.py`

rlancemartin committed:
Revert re-naming and moving of document_transformers.py

rlancemartin committed:
Revert changes to langchain/document_transformers.py

rlancemartin committed:
Need to move document_transformers.py due to redundant naming w new subdir

rlancemartin committed:
Merge branch 'master' into jasonwcfan/master

rlancemartin committed:
Update init

rlancemartin committed:
fmt

baskaryan commented on langchain/document_transformers/__init__.py:
```diff
@@ -0,0 +1,23 @@
+    "DoctranPropertyExtractor",
+    "EmbeddingsClusteringFilter",
+    "EmbeddingsRedundantFilter",
+    "_DocumentWithState",
```
no need to export private stuff

rlancemartin replied:
ya, legacy from the initial commit. pulling them out now.

jasonwcfan replied:
I did so because previously I was running into failures because there are tests that import these internal methods and classes, e.g [test_embeddings_filter.py](https://github.com/hwchase17/langchain/blob/master/tests/integration_tests/retrievers/document_compressors/test_embeddings_filter.py)

But not sure if there's another way to run the tests that makes these methods available. Should I go ahead and remove them?

rlancemartin replied:
ya, eh may need them:
```
tests/unit_tests/test_document_transformers.py:2: error: Module "langchain.document_transformers" has no attribute "_filter_similar_embeddings"  [attr-defined]
langchain/retrievers/document_compressors/embeddings_filter.py:8: error: Module "langchain.document_transformers" has no attribute "_get_embeddings_from_stateful_docs"  [attr-defined]
langchain/retrievers/document_compressors/embeddings_filter.py:8: error: Module "langchain.document_transformers" has no attribute "get_stateful_documents"  [attr-defined]
```

baskaryan replied:
we should just update those important statements to import from the specific files instead of from document_transformers init

rlancemartin replied:
ya ill clean this all up.

baskaryan commented on langchain/document_transformers/text_extract.py:
```diff
@@ -0,0 +1,42 @@
+    ) -> Sequence[Document]:
+        """Extracts properties from text documents using doctran."""
+
+        properties = kwargs.get("properties", None)
```
should this and openai_api_key just be passed in during init?

rlancemartin replied:
threw lint error from `@abstractmethod`:

```
    @abstractmethod
    async def atransform_documents(
        self, documents: Sequence[Document], **kwargs: Any
    ) -> Sequence[Document]:
```

baskaryan replied:
as in to the class `__init__` method, and accessed as instance attributes in the function

rlancemartin replied:
ya that's reasonable; none of these classes were using init methods. 

@jasonwcfan will take care of this.

baskaryan commented on langchain/document_transformers/text_translate.py:
```diff
@@ -0,0 +1,39 @@
+        ]
+        for i, doc in enumerate(doctran_docs):
+            doctran_docs[i] = await doc.translate(language=language).execute()
+        return [Document(page_content=doc.transformed_content) for doc in doctran_docs]
```
is dropping original metadata intentional?

jasonwcfan replied:
Nope, thanks for catching that. Will add the fix

rlancemartin replied:
@jasonwcfan had this initially; I'm testing add-back now.

rlancemartin committed:
Address comments

rlancemartin committed:
Fix tests

rlancemartin committed:
Upddate interface

rlancemartin committed:
fmt

rlancemartin commented:
This is good to go in once @jasonwcfan adds init methods and does a final sweep of tests.

@baskaryan minor point that Vercel deploy test is failing (non blocking).

jasonwcfan committed:
move params to init

baskaryan commented:
can fix vercel and lint issues and land from here @jasonwcfan!

baskaryan committed:
Merge branch 'master' into jasonwcfan/master

baskaryan committed:
fix

rlancemartin commented:
Thanks @baskaryan! Great. Should we wait for the Vercel fix?

baskaryan committed:
cr

baskaryan commented on docs/extras/use_cases/question_answering/chat_qa_documents_to_convert.ipynb:
@rlancemartin can this be deleted?

rlancemartin replied:
oh, ya! weird, not sure how that got added here!

baskaryan committed:
undo

baskaryan committed:
cr

baskaryan committed:
final
